### ISSUE TITLE
portaudio segfaults iff compiled

### ISSUE THREAD

**Issue by dancor** [sw17ch/portaudio#3]
For a simple sawtooth wave using the blocking IO approach, I'm getting a segfault iff I compile:
https://gist.github.com/3520964

I'm not sure if this is related to me being on a 64-bit machine?

**Comment by avaitla** [comment]
Thanks for the example. I will look into it this week.

**Comment by dancor** [comment]
Thank you!  I'm still working on my FFI savviness, I wish I could help
debug it more.

**Comment by hanshoglund** [comment]
Same problem here. Running [this](https://gist.github.com/hanshoglund/5544499) on OS X 10.7 (64-bit) crashes when compiled but works nice with `runhaskell`.

**Comment by dancor** [comment]
Still seeing the same behavior with newer Linux and GHC:
- Haskell package portaudio-0.2.3
- Ubuntu 13.04, but with 3.11 kernel, SMP, x86_64
- ghc 7.6.2
- Ubuntu package portaudio19-dev

**Comment by avaitla** [comment]
@potat0 submitted this great patch which seems to fix your issue on my machine by no longer segfaulting @dancor   https://github.com/sw17ch/portaudio/pull/5. Confirm if you have a chance.

I think there is still something wrong with the other example, which needs more investigation @hanshoglund

**Comment by dancor** [comment]
Yes it worked: on my gist above, still on 64-bit Ubuntu, the sawtooth wave plays, then the program exits with no segfault, under both compilation and runghc. Thank you very much everyone!

I am also seeing that hanshoglund's gist still segfaults under compilation only.

**Comment by raymoo** [comment]
I'm getting this error when compiling, but only when I set the suggested latency low. I was using the defaultLowOutputLatency, which was around 8ms. Here's a minimal breaking example (at least on my machine): http://lpaste.net/142590. It stops segfaulting if I have no callback, but it still crashes with this trivial callback. So maybe it has something to do with how portaudio calls back into haskell?

EDIT: I noticed this only happens for me on low latencies. If I set the latency to 0.1 or 0.05 it doesn't segfault for me.

Also forgot to mention I'm doing this on a Debian x64 system with GHC 7.10.2.

EDIT2: I don't get the issues when compiling with threaded runtime
